java/src/jmri/jmrix/can/adapters/gridconnect/GcSerialDriverAdapter.java

```diff
@@ -128,14 +133,166 @@ protected GcTrafficController makeGcTrafficController() {
+                tail.e = e;
+                if (++errorCount > MAX_IO_ERRORS_TO_ABORT) {
+                    log.error("Closing read thread due to too many IO errors", e);
+                    return null;
+                } else {
+                    log.warn("Error reading serial port " + portName, e);
+                }
+            }
+            return tail;
+        }
+
+        /**
+         * Implementation of the buffering thread.
+         */
+        private void readThreadBody() {
+            BufferEntry tail;
+            while(true) {
+                // Try to read one byte to block the thread.
+                tail = tryRead(1);
+                if (tail == null) {
+                    return;
+                }
+                // NOTE: in order to reuse this class in a generic context, we need to add support
+                // for the underlying input stream persistently returning EOF. That does not
+                // happen on a serial port.
+                if (tail.len > 0 || tail.e != null) {
+                    readAhead.add(tail);
+                } else {
+                    continue;
+                }
+                // Read as many bytes as we have in large increments. REading 128 bytes is a good
+                // compromise between throughput (4 gridconnect packets per kernel IO) but not
+                // wasting a lot of memory if less data actually shows up.
+                do {
+                    tail = tryRead(128);
+                    if (tail == null) {
+                        return;
+                    }
+                    if (tail.len > 0 || tail.e != null) {
+                        readAhead.add(tail);
+                    } else {
+                        break;
+                    }
+                } while(true);
+            }
+        }
+
+        /**
+         * We queue objects of this class between the read thread and the actual read() methods.
+         */
+        private class BufferEntry {
+            // data payload
+            byte[] data;
+            // how many bytes of data are filled in
+            int len = 0;
+            // an exception was caught reading the input stream
+            IOException e = null;
+        }
+
+        @Override
+        public int read() throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public int read(byte[] bytes) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public synchronized int read(byte[] bytes, int skip, int len) throws IOException {
+            if (skip != 0) {
+                throw new UnsupportedOperationException();
+            }
+            if (head == null || headOfs >= head.len) {
+                while (true) {
+                    try {
+                        head = readAhead.take();
+                        break;
+                    } catch (InterruptedException e) {
```

Comment:
This exception must be dealt with. See http://www.jmri.org/help/en/html/doc/Technical/Threads.shtml for instructions on how to deal with it.

Reply:
done

Reply:
The comments in your method leave it unclear if the GUI thread or the Layout thread is the correct thread. Make this method internally consistent in both its comments and its code.

I may have made a mistake specifying which thread which in the original comment, but now its unclear to anyone which thread is correct.